### PR TITLE
Make summarizer client read-only to layers below container runtime

### PR DESCRIPTION
**Pull request by agarwal-navin**
Summarizer client should never have local changes as it can lead to eventual consistency and potential other issues. In order to prevent user / app changes in the summarizer client that can lead to this, the summarizer client should be in read-only mode.

This change adds a delta manager proxy at the container runtime layer and is passed to layers below. The delta manager proxy returns read-only for summarizer clients. This will help users / app prevent making local changes in summarizer by checking read-only state.

Note that summarizer has to send summary op and upload the summary but those are are the container runtime layer which uses the actual delta manager and not the proxy.

[AB#3450](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3450)

### CODE REVIEW

File: packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	ISequencedDocumentMessage,
+	ISignalMessage,
+} from "@fluidframework/protocol-definitions";
+import { summarizerClientType } from "./summary";
+
+/**
+ * Proxy to the real IDeltaManager for the container runtime. It is used to restrict certain access to the runtime and
+ * layers below it.
+ */
+export class DeltaManagerRuntimeProxy
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts:27]: We need to give it better name. If you think it will have one responsibility, then I'd call it according to this responsibility. 
I think it's better to have many adapters like that in the future, each with its own responsibility (then merging all into one class), so I think we should rename this class / file accordingly.

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 2]
```diff
@@ -165,10 +166,6 @@ export abstract class FluidDataStoreContext
 		return this._containerRuntime.logger;
 	}
 
-	public get deltaManager(): IDeltaManager<ISequencedDocumentMessage, IDocumentMessage> {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:168]: What about ContainerRuntime.deltaManager getter? It feels like it needs to either be removed, or have a proxy object there - there should be no way to get to "real" think it seems.
If some scenario requires it, then maybe it should not be public, and appropriate comments / maybe even renames should be done to clearly articulate to callers why it's not virtualized same way as for data stores

File: packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	extends EventForwarder<IDeltaManagerEvents>
+	implements IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>
+{
+	public get IDeltaSender(): IDeltaSender {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts:31]: Can we create a base (abstract?) implementation of DM proxy, such that we do not need to reimplement these methods? And share this base class here and in the proxy that Loader uses?

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:168]: Summarizer will need access to real thing, but can we special case it somehow, and let the rest of the system (not just data stores) to work only with proxy?

File: packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts:31]: I was thinking the same thing. Let me do that.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:168]: Good point. The getter should be removed.
Also:
- We should not be passing container runtime to data store context but pass either a smaller interface or the required properties directly.
- Delta manager should not be directly exposed from data store context / data store runtime. Only the required properties from it should be exposed directly. An example of how things can go wrong - IDetlaManager exposes methods like close(), dispose() and flush() method which if called can mess things up.

File: packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaManagerRuntimeProxy.ts]: Updated the name to `DeltaManagerSummarizerProxy`

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:168]: Changed the getter to return the proxy. Added _deltaManager to ContainerRuntime that is used by that class. There are a few places in ContainerRuntime that needs to use the actual delta manager because they access readonly info. For consistency, I updated all usages in ContainerRuntime to use _deltaManager.

File: experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts [hunk 4]
```diff
@@ -193,8 +193,8 @@ export class SharedPropertyTree extends SharedObject {
 	}
 
 	private scopeFutureDeltasToPaths(paths?: string[]) {
-		const socket = (this.runtime.deltaManager as any).deltaManager.connectionManager.connection
-			.socket;
+		const socket = (this.runtime.deltaManager as any).deltaManager.deltaManager
+			.connectionManager.connection.socket;
```
Comment (agarwal-navin) [inline experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts:197]: Created https://dev.azure.com/fluidframework/internal/_workitems/edit/3553 for this.

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	public get readOnlyInfo(): ReadOnlyInfo {
+		// Summarizer clients should be read-only as far as the runtime and layers below are concerned. There shouldn't
+		// be any local changes (writes) in the summarizer except for the SummarizeOp which is generated by the runtime.
+		if (this.isSummarizerClient) {
+			return {
+				readonly: true,
+				forced: false,
+				permissions: true,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:33]: I'd leave this undefined... not sure who uses it or for what, but seems unnecessary to go so far as to say that the user has read-only permissions.

Comment (markfields) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:33]: Or maybe pass-through whatever the real DM says.

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+				storageOnly: true,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:34]: I'd pass-through whatever the real DM says for this.

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+		// except for the SummarizeOp which is generated by the runtime.
+		return this.isSummarizerClient ? false : this.deltaManager.active;
+	}
+
+	public get readOnlyInfo(): ReadOnlyInfo {
+		// Summarizer clients should be read-only as far as the runtime and layers below are concerned. There shouldn't
+		// be any local changes (writes) in the summarizer except for the SummarizeOp which is generated by the runtime.
+		if (this.isSummarizerClient) {
+			return {
+				readonly: true,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:31]: if the real DM also says `readonly: true`, maybe pass through the readonly info as-is.

File: packages/runtime/container-runtime/src/summary/summarizerTypes.ts [hunk 8]
```diff
@@ -84,7 +80,6 @@ export interface IConnectableRuntime {
 	readonly disposed: boolean;
 	readonly connected: boolean;
 	readonly clientId: string | undefined;
-	readonly deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/summary/summarizerTypes.ts:87]: Why are you changing who holds the DeltaManager?  Is it needed in more places that can't access this?

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/summarizerTypes.ts:87]: I guess it's from the observation that RunningSummarizer used to have `ISummarizerRuntime` but only ever access `deltaManager` on it - so just cut out the middle man?

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:31]: I am not sure about that. I think we need to have a defined readonly info for summarizer clients and use those. For exampe, if the main deltamanager is readonly because it was forced to, the summarizer shouldn't say so. The summarizer is read-only because it is a system (runtime) client and that is why you cannot write.

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:33]: I would say that the user does have read-only permission because it is a sytem client.

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:34]: yeah, this should be false.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -1263,7 +1276,7 @@ export class ContainerRuntime
 			this.remoteMessageProcessor.clearPartialMessagesFor(clientId);
 		});
 
-		this.summaryCollection = new SummaryCollection(this.deltaManager, this.logger);
+		this.summaryCollection = new SummaryCollection(this._deltaManager, this.logger);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1279]: It would be great to use proxy object (through getter) everywhere except where we actually need real thing. That way it will be clear where we have dependencies in summarizer on real thing.
Looking at this code (and usage of same _deltaManager on line 1152) I do not know if this is intentional, or this code does not care

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 10]
```diff
@@ -1349,11 +1365,11 @@ export class ContainerRuntime
 			}
 		}
 
-		this.deltaManager.on("readonly", (readonly: boolean) => {
+		this._deltaManager.on("readonly", (readonly: boolean) => {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1368]: Same question as above - should this be using proxy and never send ops? I.e. what is the boundary of lies?

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ * - Summarizer client should not be active to layers below the container runtime to restrict local changes.
+ */
+export class DeltaManagerSummarizerProxy
+	extends DeltaManagerProxyBase
+	implements IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>
+{
+	public get active(): boolean {
+		// Summarize clients should not be active. There shouldn't be any local changes (writes) in the summarizer
+		// except for the SummarizeOp which is generated by the runtime.
+		return this.isSummarizerClient ? false : this.deltaManager.active;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:23]: return !this.isSummarizerClient && this.deltaManager.active;

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+			};
+		}
+		return this.deltaManager.readOnlyInfo;
+	}
+
+	private readonly isSummarizerClient: boolean;
+
+	constructor(deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>) {
+		super(deltaManager);
+		this.isSummarizerClient = this.deltaManager.clientDetails.type === summarizerClientType;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:44]: I think we have a function for this test somewhere

File: packages/runtime/container-runtime/src/summary/summarizerTypes.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/summarizerTypes.ts:87]: With this change, the deltamanager on container runtime is a proxy but the summarizer classes need the real delta manager mostly because they check the `active` method on delta manager.
If these classes directly accessed container runtime's delta manager, they would get the proxy. So, I removed the direct access and passed in the real delta manager directly.

**Review by vladsud — APPROVED** [review]
Overall looks fine, but I'd like more clearly defined boundary of "lies", i.e. more invasive usage of proxy in runtime, and usage of real thing only in small number of places where we have to. I.e. not change accessing DM through getter that uses proxy.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1279]: We could do that and only pass in real delta manager where needed. This means even within container runtime, the usages will be split between the proxy and the real delta manager. I think this makes it confusing as to which one is supposed to be used where when someone wants to add a new dependency on delta manager. So, I thought we could keep it simple - withing container runtime, use the real delta manager and every where else, use the proxy.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1279]: I don't see a clean solution either way. There is no easy way to say `don't use the getter if you want the real delta manager` except for adding comments which are easy to miss.

File: packages/runtime/container-runtime/src/summary/runWhileConnectedCoordinator.ts [hunk 13]
```diff
@@ -63,13 +62,19 @@ export class RunWhileConnectedCoordinator implements ICancellableSummarizerContr
 		return this.stopDeferred.promise;
 	}
 
-	public static async create(runtime: IConnectableRuntime) {
-		const obj = new RunWhileConnectedCoordinator(runtime);
+	public static async create(
+		runtime: IConnectableRuntime,
+		deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>,
```
[the remark this answers is not in the document]
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/runWhileConnectedCoordinator.ts:67]: The problem is that this class need the real delta manager as it does the `active` check for summarization. There is no clean way for it to access the real thing if it has access to the delta manager via IConnectableRuntime.

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts (hunk 7 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:31]: Ok I like that perspective, sounds good.  If needed we can even add another field `summarizer: true | undefined` to this info object which is true here

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts (hunk 5 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:33]: Hmm I kind of see what you mean.  But the authenticated user does have permissions.  I don't think it's the same thing... and since this supports undefined that sounds best to me (unless undefined has some weird specific meaning)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 14]
```diff
@@ -930,6 +933,17 @@ export class ContainerRuntime
...
 	private readonly outbox: Outbox;
 
+	/**
+	 * The real delta manager provided by the container context. In most cases, using the proxy should be sufficient.
+	 * This should be used only if necessary.
+	 */
+	private readonly realDeltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>;
+	/**
+	 * Proxy to the real delta manager with restricted access. This is what is exposed to all usages outside
+	 * container runtime.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:943]: nit: ContainerRuntime (to clearly indicate it's the class not the layer)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 15]
```diff
@@ -930,6 +933,17 @@ export class ContainerRuntime
 	private readonly pendingStateManager: PendingStateManager;
 	private readonly outbox: Outbox;
 
+	/**
+	 * The real delta manager provided by the container context. In most cases, using the proxy should be sufficient.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:937]: Might be worth mentioning that even this may be yet another proxy.  Or just drop the word "real" from the comment.  And maybe explain a case where it's necessary to use this instead of the proxy?

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:937]: Thinking more about this, it's hard for developers working in this class to understand when to use which.  It requires knowing exactly what the proxy one overrides and why.

Part of the problem is that this class is just so huge.

Anyway, I'm brainstorming about how to make this clearer. How about this?

* rename `realDeltaManager` to `_innerDeltaManager`
* Get rid of `proxyDeltaManager` and replace `get DeltaManager` with `readonly DeltaManager` that's set in the constructor
* Update most of the usages of `this.DeltaManager` to `this._innerDeltaManager` in this class - this should be the default (opposite of how it is now).  Only use `this.DeltaManager` when passing it out to some other component (symmetry with how external components only have access to this one)

In other words, the rule becomes:  If this class's logic needs the DM, use `_innerDeltaManager`.  If you're handing the DM out to someone else in the Runtime layer, use `DeltaManager`.

We can discuss offline if you want.

File: packages/runtime/container-runtime/src/summary/summarizerTypes.ts (hunk 8 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/summary/summarizerTypes.ts:87]: Looks like it was mostly reverted - why not go all the way back?

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/summarizerTypes.ts:87]: Oh I see, never mind.  Looks good to lock it down a bit more since it's no longer used - you're passing just the `isActive` callback to that code now instead.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 15 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:937]: > Might be worth mentioning that even this may be yet another proxy. Or just drop the word "real" from the comment. And maybe explain a case where it's necessary to use this instead of the proxy?

As far as this class / layer is concerned its the real delta manager. It shouldn't care if the layers above are giving it a proxy. Same logic applies to layers below the container runtime. For them, its the real delta manager, they shouldn't care if it is proxied by upper layers.
I think renaming it something like "innerDeltaManager" is fine but just wanted to make that expectation clear.

File: packages/runtime/container-runtime/package.json [hunk 16]
```diff
@@ -120,6 +120,9 @@
 			},
 			"ClassDeclaration_ContainerRuntime": {
 				"forwardCompat": false
+			},
+			"InterfaceDeclaration_IConnectableRuntime": {
+				"backCompat": false
```
Comment (markfields) [inline packages/runtime/container-runtime/package.json:125]: This is a breaking change.  I think you just have to mark it as deprecated.  Probably mark the whole `IConnectableRuntime` interface as deprecated and remove it from exports in `next`, at least it's not used in Bohemia codebase.

File: experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts (hunk 4 above)
Comment (markfields) [reply experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts:197]: What about backcompat.  Don't you need to check both depth levels?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 15 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:937]: As for your proposal, I had it that way too, i.e., all usages in ContainerRuntime uses realDeltaManager and the getter returns a proxy. However, Vlad had a good point here - https://github.com/microsoft/FluidFramework/pull/14354#discussion_r1122013108. Basically, only use the real delta manager if you know what you are doing. I don't see a clean solution either way since ContainerRuntime needs to expose deltaManager.
As fas the developers guidance, they should use ContainerRuntime::deltaManager by default. Only if they really need it and care, they should use the real delta manager.

File: experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts (hunk 4 above)
Comment (markfields) [reply experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts:197]: `while(deltaManager.deltaManager !== undefined) { deltaManager = deltaManager.deltaManager; }`  😹

File: packages/runtime/container-runtime/package.json (hunk 16 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/package.json:125]: I tried removing it but it adds a "missing export" warning because ContainerRuntime implements ISummarizerRuntime and ContainerRuntime is exported.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 15 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:937]: Discussed offline.  No good solution here, opting for sprinkling some more comments around and hoping for the best.

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:44]: I could not find one :(

File: packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaManagerSummarizerProxy.ts:33]: That makes sense. Changing it to undefined.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 15 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Updated names and comments. Take a look now.

**Review by markfields — APPROVED** [review]
🎂
